extensions/amp-pan-zoom/1.0/component-ts.tsx

```diff
@@ -0,0 +1,267 @@
+import {
+  GestureHandlers,
+  UserGestureConfig,
+  UserHandlers,
+  useGesture,
+} from '@use-gesture/react';
+import type {ComponentChildren, Ref} from 'preact';
+
+import {scale as cssScale, px, translate} from '#core/dom/style';
+
+import * as Preact from '#preact';
+import {useImperativeHandle, useRef} from '#preact';
+import {forwardRef} from '#preact/compat';
+import {ContainWrapper} from '#preact/component';
+import {useResizeObserver} from '#preact/hooks/useResizeObserver';
+
+import {useStyles} from './component.jss';
+import {usePanZoomState} from './hooks/usePanZoomState';
+
+const DOUBLE_TAP_TIME = 500; // Maximum time between tap starts
+
+const gestureConfig: Omit<UserGestureConfig, 'target'> = {
+  drag: {
+    filterTaps: true,
+    tapsThreshold: 10,
+  },
+};
+
+export type BentoPanZoomProps = {
+  children?: ComponentChildren;
+  controls?: boolean;
+  initialScale?: number | string;
+  initialX?: number | string;
+  initialY?: number | string;
+  maxScale?: number | string;
+};
+
+export type BentoPanZoomApi = {
+  transform(scale: number, x: number, y: number): void;
+};
+
+function getElementPosition(
+  clientX: number,
+  clientY: number,
+  element: HTMLElement
+) {
+  const elBounds = element./* REVIEW */ getBoundingClientRect();
```

Comment:
This causes a full CSS recalc. Is it possible to use an async source (eg, InOb) to get this? Especially during a drag operation, this may cause a noticeable frame drop.

Reply:
My understanding is that `getBoundingClientRect` causes a reflow, which means it could be a performance issue if the browser hasn't yet calculated the layout.
However, this is only being used during an event handler (`onDrag` / `onPinch`), when the browser has already performed a full layout and does not need to reflow.  This is the ideal time to retrieve this bounds, since we need to know the element's exact `x/y` offsets (determined by window size, scroll).